### ISSUE TITLE
module_utils.basic.is_executable returns true for directories

### ISSUE THREAD

**Issue by haldean** [ansible/ansible#18688]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
module_utils

##### ANSIBLE VERSION
current devel (https://github.com/ansible/ansible/commit/461dcf8ce3edb0fafcbb608c846b9a68eddde019)

##### CONFIGURATION
(unrelated)

##### OS / ENVIRONMENT
POSIX environments that use PATH (Linux, OSX, *BSD, etc)

##### SUMMARY
The [`is_executable` function](https://github.com/ansible/ansible/blob/461dcf8ce3edb0fafcbb608c846b9a68eddde019/lib/ansible/module_utils/basic.py#L557) in module_utils.basic doesn't check whether the input file is a directory or not. This causes [`get_bin_path`](https://github.com/ansible/ansible/blob/461dcf8ce3edb0fafcbb608c846b9a68eddde019/lib/ansible/module_utils/basic.py#L1815) in the same module, which is used by a number of ansible modules, to return directories as executables instead of continuing the search to find the actual, appropriate executable. This manifested itself for me by causing `unarchive` to fail when it found a directory named `tar` on my `PATH` before the actual directory named `tar`. `bash` and other shells skip directories when searching on `PATH`.

Note that the executable bit on directories are used to signal that the user can cwd into the directory; it has nothing to do with actually being executable.

If the correct fix is just to add a `not os.path.isdir` to `is_executable`, let me know and I'd be happy to send a PR! I'm not familiar enough with the entire codebase to know whether that's an acceptable fix, though.

**Comment by abadger** [comment]
Yeah, either adding ```not os.path.isdir()``` or ```os.path.isfile()``` (I haven't looked into whether the latter will deal with links correctly or if there's other cornercases).  Modify the docstring a little bit to the docstring to make it explicit that is_executable() is for finding an executable file and will not find a directory.  right now it's implicit in the docstring but it's not clear whether that's just the way it was worded or if it is actually the case  (I looked at all the callers of is_executable() and it seems that indeed, everything that calls it is expecting to find an executable file rather than something with the executable bit set.).

**Comment by Akasurde** [comment]
resolved_by_pr #34311

**Comment by abadger** [comment]
Is that the right fix or is the right fix to make get_bin_path check for
both executable and directory?

**Comment by Akasurde** [comment]
@abadger This fix is only for is_executable function. get_bin_path is already patched to ignore directories in user path - [here](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/basic.py#L2203) as discussed in issue.

**Comment by abadger** [comment]
Okay, so I read the function and that prodded my memory:

> * Most times we really want to know "Can the current user execute this
file" This function does not tell us that.  Only if an execute bit is set.

I remember writing that after having a discussion with Jimi and bcoca about
whether to change that function.  We decided not to change for backwards
compatibility concerns. Calling code could be looking for whether a file
has executable bits  just as easily as it could be looking for whether a
file is a program.  We decided to document what this function is returning
so that people would know what to expect it to return instead of changing
it and breaking that use case.

We could add a new kwarg to enable that use case if you have some other
code that needs it.  However, I think that we might have decided not to do
that because there were other wrinkles.  (Whether the file was executable
by this specific user?  Fsacls are mentioned as a shortcoming in the
function's docstring, for instance).  So I'm not sure we should do that
without adding handling for the other cases as well.
